File: backend/app/schemas/bracelet_alerts.py
"""API-схемы мониторинга браслетов."""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class VitalAlertView(BaseModel):
    metric: str
    label: str
    value: float
    unit: str
    level: str
    message: str
    normal_range: str


class MetricThresholdValues(BaseModel):
    label: str
    unit: str
    normal_min: Optional[float] = None
    normal_max: Optional[float] = None
    warning_low: Optional[float] = None
    warning_high: Optional[float] = None
    critical_low: Optional[float] = None
    critical_high: Optional[float] = None


class MetricThresholdOverrideInput(BaseModel):
    normal_min: Optional[float] = None
    normal_max: Optional[float] = None
    warning_low: Optional[float] = None
    warning_high: Optional[float] = None
    critical_low: Optional[float] = None
    critical_high: Optional[float] = None


class PatientVitalThresholdsUpdate(BaseModel):
    """Произвольный набор канонических показателей (pulse, spo2, temp, …)."""
    pulse: Optional[MetricThresholdOverrideInput] = None
    spo2: Optional[MetricThresholdOverrideInput] = None
    temp: Optional[MetricThresholdOverrideInput] = None
    respiration: Optional[MetricThresholdOverrideInput] = None
    press: Optional[MetricThresholdOverrideInput] = None
    hrv: Optional[MetricThresholdOverrideInput] = None
    stress: Optional[MetricThresholdOverrideInput] = None
    sleep: Optional[MetricThresholdOverrideInput] = None
    battery: Optional[MetricThresholdOverrideInput] = None


class PatientVitalThresholdsResponse(BaseModel):
    patient_id: int
    has_custom: bool
    defaults: Dict[str, MetricThresholdValues]
    overrides: Optional[Dict[str, Dict[str, float]]] = None
    effective: Dict[str, MetricThresholdValues]


class PatientBraceletView(BaseModel):
    patient_id: int
    patient_name: str
    ble_mac: Optional[str] = None
    room_number: Optional[str] = None
    bed_number: Optional[str] = None
    admission_date: Optional[str] = None
    online: Optional[bool] = None
    metrics: Dict[str, Any] = Field(default_factory=dict)
    alerts: List[VitalAlertView] = Field(default_factory=list)
    has_custom_thresholds: bool = False


class UnassignedBleDeviceView(BaseModel):
    mac: str
    online: Optional[bool] = None
    metrics: Dict[str, Any] = Field(default_factory=dict)


class BraceletAssignmentPair(BaseModel):
    patient_id: int
    patient_name: str
    ble_mac: str
    room_number: Optional[str] = None
    bed_number: Optional[str] = None


class AssignBraceletRequest(BaseModel):
    patient_id: int
    ble_mac: str


class AssignBraceletResponse(BaseModel):
    pair: BraceletAssignmentPair
    monitoring_connected: bool
    message: str
    error: Optional[str] = None


class UnassignBraceletResponse(BaseModel):
    patient_id: int
    patient_name: str
    ble_mac: str
    message: str


class DistributeBraceletsResponse(BaseModel):
    assigned_count: int
    pairs: List[BraceletAssignmentPair] = Field(default_factory=list)
    patients_without_mac_remaining: int
    unassigned_devices_remaining: List[str] = Field(default_factory=list)
    monitoring_connected: bool
    message: str
    error: Optional[str] = None


class BraceletOverviewResponse(BaseModel):
    checked_at: str
    patients_total: int
    patients_with_ble: int
    patients_online: int
    patients_without_mac: int = 0
    alerts_found: int
    monitoring_connected: bool
    max_bot_configured: bool
    alerts_enabled: bool
    error: Optional[str] = None
    patients: List[PatientBraceletView] = Field(default_factory=list)
    unassigned_devices: List[UnassignedBleDeviceView] = Field(default_factory=list)


class BraceletCheckResponse(BaseModel):
    checked_at: str
    alerts_found: int
    alerts_sent: int
    alerts_skipped_dedup: int
    monitoring_connected: bool
    message: str

File: frontend/src/types/braceletAlerts.ts
export type VitalAlertLevel = 'normal' | 'warning' | 'critical';

export interface VitalAlert {
  metric: string;
  label: string;
  value: number;
  unit: string;
  level: VitalAlertLevel;
  message: string;
  normal_range: string;
}

export interface MetricThresholdValues {
  label: string;
  unit: string;
  normal_min?: number | null;
  normal_max?: number | null;
  warning_low?: number | null;
  warning_high?: number | null;
  critical_low?: number | null;
  critical_high?: number | null;
}

export interface PatientVitalThresholds {
  patient_id: number;
  has_custom: boolean;
  defaults: Record<string, MetricThresholdValues>;
  overrides: Record<string, Record<string, number>> | null;
  effective: Record<string, MetricThresholdValues>;
}

export interface PatientBraceletStatus {
  patient_id: number;
  patient_name: string;
  ble_mac?: string | null;
  room_number?: string | null;
  bed_number?: string | null;
  admission_date?: string | null;
  online?: boolean | null;
  metrics: Record<string, number | string | boolean | null>;
  alerts: VitalAlert[];
  has_custom_thresholds?: boolean;
}

export interface UnassignedBleDevice {
  mac: string;
  online?: boolean | null;
  metrics: Record<string, number | string | boolean | null>;
}

export interface BraceletAssignmentPair {
  patient_id: number;
  patient_name: string;
  ble_mac: string;
  room_number?: string | null;
  bed_number?: string | null;
}

export interface UnassignBraceletResult {
  patient_id: number;
  patient_name: string;
  ble_mac: string;
  message: string;
}

export interface AssignBraceletResult {
  pair: BraceletAssignmentPair;
  monitoring_connected: boolean;
  message: string;
  error?: string | null;
}

export interface DistributeBraceletsResult {
  assigned_count: number;
  pairs: BraceletAssignmentPair[];
  patients_without_mac_remaining: number;
  unassigned_devices_remaining: string[];
  monitoring_connected: boolean;
  message: string;
  error?: string | null;
}

export interface BraceletOverview {
  checked_at: string;
  patients_total: number;
  patients_with_ble: number;
  patients_online: number;
  patients_without_mac?: number;
  alerts_found: number;
  monitoring_connected: boolean;
  max_bot_configured: boolean;
  alerts_enabled: boolean;
  error?: string | null;
  patients: PatientBraceletStatus[];
  unassigned_devices?: UnassignedBleDevice[];
}

export interface BraceletCheckResult {
  checked_at: string;
  alerts_found: number;
  alerts_sent: number;
  alerts_skipped_dedup: number;
  monitoring_connected: boolean;
  message: string;
}
